src/viper/trans.ml

```diff
@@ -356,21 +356,21 @@ and dec_field' ctxt d =
       { ctxt with ids = Env.add f.it (Method, note) ctxt.ids },
       None, (* no perm *)
       None, (* no init *)
-      fun ctxt' ->
+      fun ctxt ->
         let open Either in
         let self_id = !!! (Source.no_region) "$Self" in
         let method_args = args p in
-        let method_args' = List.map (fun (id, t) -> id, tr_typ ctxt' t) method_args in
-        let ctxt'' = { ctxt'
+        let method_args' = List.map (fun (id, t) -> id, tr_typ ctxt t) method_args in
+        let ctxt = { ctxt
           with self = Some self_id.it;
                ids = List.fold_left (fun env (id, t) -> Env.add id.it (Local, t) env) ctxt.ids method_args }
```

Comment:
here semantic also have been changed.
Before we used global `ctxt` that comes from argument of `dec_field'` while and now you are using `ctxt` from lambda.
Are you sure that it is legit assuption?

Reply:
> Are you sure that it is legit assuption?

Yeah, actually it was the buggy place. AFAIU the context in the lambda is the one with all top-level definitions. As I see, Motoko doesn't support duplicate top-level names and all top-levels could be accessed in every method

Reply:
Actually I've understood what is happened:
Problem is that monorphisation is changing order of declarations:
```
xarray: [var Nat]
func copy_xarray;
func reverseArray<T>;
```
becomes
```
func reverseArray$Nat; // here there is no xarray
xarray: [var Nat]
func copy_xarray;
```

In the going "top-level"/"declaration" context there is no "xarray" if function is polymorphic because it goes later while in the "body" context (after handling all declarations) it exists.

You are saying that using both contexts in translation of function body was a mistake. Probably very old one. Okay I agree.
I guess main idea behind this code was is to have 1-phase translation handling declarations and definitions at ones.